### PR TITLE
sql: add delay before releasing schema change lease on failure

### PR DESCRIPTION
**Pull request by dt**
In the last release, we added an eager release of the schema change lease after an error.
    This was done to allow another node to retry sooner, without having to wait for the lease to expire,
    for example if the executing node hit some transient error.

    This change blocks for some delay before a schema changer releases its
    lease after a failure to put a upper bound on the rate at which a
    descriptor lease might be acquired/released.

    While a failure should ideally not happen in the first place or be handled if
    it does, if a failure does occur, delaying any reacquisition of the lease by
    this node or any other node for some some short duration achieves two things:
    1) provides chance for any transient issues to clear up or a node to recover
    2) more importantly, it creates an upper-bound on the rate at which a given
    descriptor's lease can be acquired and released, regardless of what other
    failures may or may not occur.

    The later upper-bound can help ensure that, for example, a runaway retry loop
    that fails to recognize a irrecoverable error does not end up in a tight loop
    that just thrashes the descriptor's lease (leading to other problems).

### CODE REVIEW

**Review by andreimatei — COMMENTED** [review]
Was there a specific case for which we added this releasing of the lease? It seems to me like if the node is able to 
If we don't release the lease, what governs the retry speed?

I'm not very inspired by this patch. I say we stop retrying schema changes. You've never expressed an opinion on that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38219#-:-Lhad3vH8A9HA1_VdGIO:b-3fpj2h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dt, @jordanlewis, and @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38219#-LhadFi07Q7DGQ7gS4Hs:-LhadFi07Q7DGQ7gS4Ht:b-pcn8dd) ([raw file](https://github.com/cockroachdb/cockroach/blob/35c6e051b42209f663ecb9a537e9e5a88a3079a5/pkg/sql/schema_changer.go#L65)):*
> ```Go
> var schemaChangeFailureReleaseDelay = settings.RegisterNonNegativeDurationSetting(
> 	"schemachanger.lease.failure_release_delay",
> 	"amount of time to delay releasing a schema change lease after a failure",
> ```

let's make this cluster setting "hidden" if it's not already
___
*[pkg/sql/schema_changer.go, line 572 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38219#-LhadjOl0XK-kTYs0p9r:-LhadjOl0XK-kTYs0p9s:b-2g90bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/35c6e051b42209f663ecb9a537e9e5a88a3079a5/pkg/sql/schema_changer.go#L572)):*
> ```Go
> }
> 
> func (sc *SchemaChanger) failureReleaseDelay(ctx context.Context) {
> ```

this needs a good comment
___
*[pkg/sql/schema_changer.go, line 573 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38219#-Lhact6DBCTrptUeVLT7:-Lhact6DBCTrptUeVLT8:b14zhi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/35c6e051b42209f663ecb9a537e9e5a88a3079a5/pkg/sql/schema_changer.go#L573)):*
> ```Go
> 
> func (sc *SchemaChanger) failureReleaseDelay(ctx context.Context) {
> 	return
> ```

?
Add a test?

**Comment by dt** [comment]
@andreimatei I updated the desc with a little more of the reasoning for why I think this change a *part* of a more defensive approach to failures. I don't think anyone is arguing that we shouldn't fix any other failure-causing bugs and correctly fail completely on fatal ones, but if/when failures are mis-diagnosed as retryable (or even if legitimately retryable error is encountered), we still don't want this to be able to happen in such a tight loop.

File: pkg/sql/schema_changer.go [hunk 1]
```diff
@@ -562,6 +569,32 @@ func (sc *SchemaChanger) maybeDropTable(
 	return nil
 }
 
+// failureReleaseDelay blocks for some delay, and should be to be called before
```
Comment (andreimatei) [inline pkg/sql/schema_changer.go]: should be to be

**Comment by andreimatei** [comment]
> I don't think anyone is arguing that we shouldn't fix any other failure-causing bugs and correctly fail completely on fatal ones

Well right now we retry on pretty much any error. So it's not that we have a bug that's causing us to retry (well, we also have a bug), but that we behave badly by design. So that's why I'd start there. Cause if we did something about that it'd be pretty unlikely that we'd need more defenses sprinkled in random places.

And then generally I think I wouldn't retry a schema change indefinitely regardless of the error it's getting - even on the most retriable error. After some time, give up and ask the administrator to press play again when they thing conditions have improved. So if you want defensive coding, I'd do that.
We have infinite retries in other places but fewer and fewer of them. And the ones that we do have have some escape hatch for stopping them (e.g. for SQL retries you can close the connection or set a timeout beforehand).
Here there's `CANCEL JOB` which I hope would generally stop a schema change - except presumably not in the case that's triggered this PR where the job actually succeeded. But that feels different - I barely know about this mechanism.

Finally, if you insist on doing some throttling, I'd throttle the execution of the schema change, not the releasing of this lease. I think it's more natural to tell the schema change scheduler "don't run this schema change for the next 10 seconds" (and thus don't acquire a lease) than to block some random goroutine to pretend that we're still running a schema change when we're not.
In other words, I see this lease acquisition/release as incidental to everything. I wouldn't focus on it or protect at this level. Cause like if you want to add defenses at such a low level, then why stop at the lease? Why not throttle any writes to any descriptor? Or to any key in the system config range for that matter? (not that I'd encourage that)
Except I'd focus on the lease to move it the hell out of the descriptor cause it just doesn't belong there - although I think @jordanlewis has tried and seemed pretty discouraged.

If you convince others though, I don't care that much about this patch.

**Review by andreimatei — CHANGES_REQUESTED** [review]

**Comment by andreimatei** [comment]
But you haven't really answered my question from before: why do we even release the lease if we want a retry to happen? I understand that releasing the lease allows another node to take it (probabilistically?) but why is that a good thing? Why would we want another node to retry instead of the current leaseholder to retry? If we're going to retry, isn't the current leaseholder in a better position to do it? Unless it's partitioned, but if it is the lease release probably won't work anyway.

**Comment by dt** [comment]
I think we'll just move to jobs-based leases at some point so I think I'd rather not mess with this.
